### ISSUE TITLE
Firebase Cloud Messaging does not Work with SPM

### ISSUE THREAD

**Issue by cyrussinc** [firebase/firebase-ios-sdk#7438]
### Step 0: Are you in the right place?

  * For issues or feature requests related to __the code in this repository__
    file a Github issue.
    * If this is a __feature request__ please use the Feature Request template.
  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the `firebase` tag.
  * For general (non-iOS) Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For backend issues, console issues, and other non-SDK help that does not fall under one
    of the above categories, reach out to
    [Firebase Support](https://firebase.google.com/support/).
  * Once you've read this section and determined that your issue is appropriate for
    this repository, please delete this section.

### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12
  * Firebase SDK version: 7.5.4
  * Installation method: `CocoaPods | Carthage | Zip file | **Swift Package Manager**` (select one)
  * Firebase Component: _____ (Auth, Core, Database, Firestore, Messaging, Storage, etc)

### [REQUIRED] Step 2: Describe the problem
I am attempting to implement push notifications w/cloud messaging. After following this tutorial: https://www.youtube.com/watch?v=vvq0etotS8M I am unable to successfully receive notifications. I already tried with Cocoapods v 6, and that worked fine. Does this have to do with the use of SPM instead of the traditional Cocoapods?
#### Steps to reproduce:

What happened? How can we make the problem occur?
This could be a description, log/console output, etc.

If you have a downloadable sample project that reproduces the bug you're reporting, you will
likely receive a faster response on your issue.

#### Relevant Code:

```
// TODO(you): code here to reproduce the problem
Get SPM versions of FirebaseMessaging and FirebaseAnalytics
Follow the linked tutorial (just a traceover the firebase ios cloud messaging docs).

```

**Comment by paulb777** [comment]
@cyrussinc Sorry about the trouble. There should be nothing about SPM that impacts FirebaseMessaging's usability. We'll likely need more detail and/or a reproducible example to provide additional help.

Also, there is no version 7.5.4.  Do you mean 7.5.1?

**Comment by cyrussinc** [comment]
I apologize; I meant to cite Version 7.5.1. I can provide the AppDelegate file with the Firebase Code. 

As for Certificates, Keys and Identifiers, I am using an APNs Authentication Key from my Apple Developer account, and have obtained a certificate with the associated Bundle ID with Push Notifications enabled.

I am almost certain the issue lies in the client side code.
 
Below is my AppDelegate file.
```
//  AppDelegate.swift
//  FirebaseMessagingConfiguration
//
//
//Using SPM Firebase Library Version 7.5.1

import UIKit
import CoreData
import FirebaseMessaging
import Firebase


@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
    
    
    var window: UIWindow?
    
    let gcmMessageIDKey = "gcm.Message_ID"
    
    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey : Any] = [:]) -> Bool {
        print(url)
        return true
    }
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        if #available(iOS 10.0, *) {
            // For iOS 10 display notification (sent via APNS)
            UNUserNotificationCenter.current().delegate = self
            
            let authOptions: UNAuthorizationOptions = [.alert, .badge, .sound]
            UNUserNotificationCenter.current().requestAuthorization(
                options: authOptions,
                completionHandler: {_, _ in })
        } else {
            let settings: UIUserNotificationSettings =
                UIUserNotificationSettings(types: [.alert, .badge, .sound], categories: nil)
            application.registerUserNotificationSettings(settings)
        }
        
        application.registerForRemoteNotifications()
        
        
        FirebaseApp.configure()
        Messaging.messaging().delegate = self
        //  Messaging.messaging().shouldEstablishDirectChannel = true
        
        
        return true
    }
    
    
    func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable: Any],
                     fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
        if let messageID = userInfo[gcmMessageIDKey] {
            print("Message ID: \(messageID)")
        }
        
        // Print full message.
        print(userInfo)
        
        completionHandler(UIBackgroundFetchResult.newData)
    }
}


//MARK: Firebase Messaging Extensions

extension AppDelegate: MessagingDelegate{
    func messaging(_ messaging: Messaging, didReceiveRegistrationToken fcmToken: String?) {
        print("Firebase registration token: \(String(describing: fcmToken))")
        
        let dataDict:[String: String] = ["token": fcmToken ?? ""]
        
        
        NotificationCenter.default.post(name: Notification.Name("FCMToken"), object: nil, userInfo: dataDict)
        // TODO: If necessary send token to application server.
        // Note: This callback is fired at each app startup and whenever a new token is generated.
    }
   
}


@available(iOS 10, *)
extension AppDelegate : UNUserNotificationCenterDelegate {
    
    // Receive displayed notifications for iOS 10 devices.
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                willPresent notification: UNNotification,
                                withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        let userInfo = notification.request.content.userInfo
        
        // With swizzling disabled you must let Messaging know about the message, for Analytics
        // Messaging.messaging().appDidReceiveMessage(userInfo)
        
        // Print message ID.
        if let messageID = userInfo[gcmMessageIDKey] {
            print("Message ID: \(messageID)")
        }
        
        // Print full message.
        print(userInfo)
        
        // Change this to your preferred presentation option
        completionHandler([[.alert, .sound]])
    }
    
    func userNotificationCenter(_ center: UNUserNotificationCenter,
                                didReceive response: UNNotificationResponse,
                                withCompletionHandler completionHandler: @escaping () -> Void) {
        let userInfo = response.notification.request.content.userInfo
        // Print message ID.
        if let messageID = userInfo[gcmMessageIDKey] {
            print("Message ID: \(messageID)")
        }
        
        // With swizzling disabled you must let Messaging know about the message, for Analytics
        // Messaging.messaging().appDidReceiveMessage(userInfo)
        
        // Print full message.
        print(userInfo)
        
        completionHandler()
    }
}

```

**Comment by cyrussinc** [comment]
Update: Using this AppDelegate.swift file in a Cocoapods project with Firebase Messaging 7.5.0 resulted in successful callback of the didReceive remote notification delegate method, and thus the display of an on screen notification in both background and foreground modes. 

This issue seems to be related to SPM.

**Comment by paulb777** [comment]
@morganchen12 @chliangGoogle Any ideas about what about SPM could make a difference in Messaging functionality?

**Comment by chliangGoogle** [comment]
Sorry for the delay. I'm NOT able to reproduce in SPM.  I don't think it's related to SPM as notification is delivered for me. Can you share with us the debugLog?

**Comment by chliangGoogle** [comment]
@cyrussinc In case it helps, we have a very simple test app here is the one I tested and worked for me: https://github.com/firebase/firebase-ios-sdk/tree/master/FirebaseMessaging/Apps/Sample

**Comment by cyrussinc** [comment]
@chliangGoogle how can I access the debug log? Also, I'll try out this sample app and see if I can get it setup and working.

**Comment by chliangGoogle** [comment]
To enable debug logging set the following application argument: -FIRDebugEnabled (see http://goo.gl/RfcP7r)

**Comment by cyrussinc** [comment]
I found this strange: I tested out the Cocoapods application (which works fine) on Xcode Simulator, and recieved this error message:

`[Firebase/Messaging][I-FCM012002] Error in application:didFailToRegisterForRemoteNotificationsWithError: remote notifications are not supported in the simulator`

This is FirebaseMessaging preemptively notifying me that the notifications aren't available, because it's not possible to receive notifications on IOS Simulator. I get no such message when running the SPM Application.

Is this a configuration issue with FirebaseMessaging (perhaps on my part)?
